lib/paper_trail/serializer.ex

```diff
@@ -0,0 +1,26 @@
+defmodule PaperTrail.Serializer do
+  @moduledoc false
+
+  def serialize(%Ecto.Changeset{data: data}, :insert), do: serialize(data, :insert)
+  def serialize(%Ecto.Changeset{data: data}, :delete), do: serialize(data, :delete)
+  def serialize(%Ecto.Changeset{changes: changes}, :update), do: serialize(changes, :update)
+
+  def serialize(%struct{} = model, mode) do
+    if function_exported?(struct, :__schema__, 1) do
+      relationships = struct.__schema__(:associations)
+
+      model
+      |> Map.from_struct()
+      |> Map.drop([:__meta__ | relationships])
+      |> Map.new(&{elem(&1, 0), serialize(elem(&1, 1), mode)})
```

Comment:
I find the this line and line 21 & 23 very hard to read compared to previous changes on `lib/paper_trail/multi.ex#228-231`. First it looked very clean but once one reads the logic, there is too much recursion and Im bit concerned that this might have a significant negative performance impact. 

Cant we figure out a different & better way to achieve this? Previous serialize was very straight forward, Im not even sure why we have `mode` parameter in these functions in the first place. I'd appreciate if you could describe these three lines here, and elaborate the intention.

Reply:
@izelnakri The recursion is needed to find the nested embedded models. I don’t think there‘s a way to achieve the functionality without recursion. Do you see another way?

Mode was added so that insert & delete have a full copy and update only has changes.